### ISSUE TITLE
mamba init and `fish` shell

### ISSUE THREAD

**Issue by atrejut** [mamba-org/mamba#1718]
Hi All,

I'm using fish shell, and can't seem to get `mamba activate/deactivate` to work.
`conda init` also doesn't modify fish config files by default, but it's possible to add a line to my `config.fish` script:
`eval /Users/_REDACTED_/.mambaforge/bin/conda "shell.fish" "hook" $argv | source`

with this I think I'm using the correct conda for activating/deactivating environments, but it would be nice if I could use `mamba` throughout.

Thanks,

Atreju

**Comment by Paradoxdruid** [comment]
To my understanding:
From inside your fish shell, run `bash` to open a bash shell inside fish, then run `conda init fish` to configure conda in your fish shell.  Restart your fish shell and it should work.

mamba advises that you should still use `conda activate` , NOT `mamba activate` ( https://mamba.readthedocs.io/en/latest/user_guide/mamba.html )

**Comment by atrejut** [comment]
Hi @Paradoxdruid,

yes, running `conda init fish` adds the line I already mentioned above to my `config.fish`. (It seems I don't need a bash shell, and can run that line directly from fish)

Thanks for the link to the documentation, do you have any details on /why/ `conda activate` is recommended over `mamba activate`, and whether that's expected to change in the future?

In either case, `conda activate` works fine -- but `mamba activate` results in
```
Run 'mamba init' to be able to run mamba activate/deactivate
and start a new shell session. Or use conda to activate/deactivate.
```
but running `mamba init` makes no difference whatsoever. So at least that message seems wrong on `fish`...?

**Comment by dlukes** [comment]
What `mamba init` does for Bash is to source this: https://github.com/mamba-org/mamba/blob/main/mamba/mamba/shell_templates/mamba.sh

That script basically just sets up a `mamba` shell function which sometimes delegates to `mamba` and sometimes to `conda`. Based on that, and on the `conda` function printed by `conda shell.fish hook`, I think the corresponding `mamba` function should look something like this:

```fish
# Adapted from `function conda` printed by `conda shell.fish hook`, following
# `mamba.sh`.
function mamba --inherit-variable CONDA_EXE
  if [ (count $argv) -lt 1 ]
    $CONDA_EXE
  else
    set -l mamba_exe (dirname $CONDA_EXE)/mamba
    set -l cmd $argv[1]
    set -e argv[1]
    switch $cmd
      case activate deactivate
        eval ($CONDA_EXE shell.fish $cmd $argv)
      case install update upgrade remove uninstall
        $mamba_exe $cmd $argv
        and eval ($CONDA_EXE shell.fish reactivate)
      case '*'
        $mamba_exe $cmd $argv
    end
  end
end
```

You can put that in your own Fish config, until Mamba ships its own version of this and makes sure that `mamba init fish` adds code to source it.

**Comment by wolfv** [comment]
@dlukes do you want to make a PR with that script? Would be super appreciated!

**Comment by dlukes** [comment]
Sure, happy to! I'll dogfood it for a bit first to see if I haven't messed up anywhere, but hopefully not, it was fairly straightforward, there wasn't much to change :) If @atrejut could test it out as well, it would be great!

**Comment by dlukes** [comment]
> see if I haven't messed up anywhere, but hopefully not

... and as I post this, I notice at least one bit that's wrong (without args, `$mamba_exe` should be invoked, not `$conda_exe`), so here's an updated version:

```fish
function mamba --inherit-variable CONDA_EXE
  set -l mamba_exe (dirname $CONDA_EXE)/mamba
  if [ (count $argv) -lt 1 ]
    $mamba_exe
  else
    set -l cmd $argv[1]
    set -e argv[1]
    switch $cmd
      case activate deactivate
        eval ($CONDA_EXE shell.fish $cmd $argv)
      case install update upgrade remove uninstall
        $mamba_exe $cmd $argv
        and eval ($CONDA_EXE shell.fish reactivate)
      case '*'
        $mamba_exe $cmd $argv
    end
  end
end
```

**Comment by jonashaag** [comment]
See also https://github.com/mamba-org/mamba/blob/main/libmamba/data/mamba.fish#L57 which is directly from Conda

**Comment by dlukes** [comment]
> which is directly from Conda

The one I posted is too, in the sense that it's just lightly tweaked from what `conda shell.fish hook` outputs as the `conda` function. And they're both very similar anyway. But it probably makes sense to use the one you posted as the basis, since it's already used for micromamba within this repo.

Would you like to take care of this then, @jonashaag, since you seem to have been previously involved with Fish support in Mamba? Or should I still go forward with a PR?

**Comment by jonashaag** [comment]
Not really, I did it because it was buggy but I‘d be glad not to have to work on it a lot :)

**Comment by dlukes** [comment]
Noted :) See https://github.com/mamba-org/mamba/pull/2006.

**Comment by wolfv** [comment]
Will be released with the next release! Thanks again @dlukes

**Comment by psobolewskiPhD** [comment]
I updated my mamba from 0.27 to 1.0.0
i ran `mamba init fish`
```
no change     /Users/piotrsobolewski/Dev/miniforge3/condabin/conda
no change     /Users/piotrsobolewski/Dev/miniforge3/bin/conda
no change     /Users/piotrsobolewski/Dev/miniforge3/bin/conda-env
no change     /Users/piotrsobolewski/Dev/miniforge3/bin/activate
no change     /Users/piotrsobolewski/Dev/miniforge3/bin/deactivate
no change     /Users/piotrsobolewski/Dev/miniforge3/etc/profile.d/conda.sh
no change     /Users/piotrsobolewski/Dev/miniforge3/etc/fish/conf.d/conda.fish
no change     /Users/piotrsobolewski/Dev/miniforge3/shell/condabin/Conda.psm1
no change     /Users/piotrsobolewski/Dev/miniforge3/shell/condabin/conda-hook.ps1
no change     /Users/piotrsobolewski/Dev/miniforge3/lib/python3.10/site-packages/xontrib/conda.xsh
no change     /Users/piotrsobolewski/Dev/miniforge3/etc/profile.d/conda.csh
modified      /Users/piotrsobolewski/.config/fish/config.fish

==> For changes to take effect, close and re-open your current shell. <==

Added mamba to /Users/piotrsobolewski/.config/fish/config.fish

==> For changes to take effect, close and re-open your current shell. <==
```
My ~/.config/fish/config.fish now has:
```
   1   │ 
   2   │ # >>> conda initialize >>>
   3   │ # !! Contents within this block are managed by 'conda init' !!
   4   │ eval /Users/piotrsobolewski/Dev/miniforge3/bin/conda "shell.fish" "hook
       │ " $argv | source
   5   │ 
   6   │ if test -f "/Users/piotrsobolewski/Dev/miniforge3/etc/fish/conf.d/mamba
       │ .fish"
   7   │     source "/Users/piotrsobolewski/Dev/miniforge3/etc/fish/conf.d/mamba
       │ .fish"
   8   │ end
   9   │ # <<< conda initialize <<<
```

So it looks like mamba has been added, but when I close and reopen my Terminal (macOS 12.6) 
and try `mamba activate test-env` I get
`CommandNotFoundError: No command 'conda shell'.`

`mamba info` does work:
```
     active environment : base
    active env location : /Users/piotrsobolewski/Dev/miniforge3
            shell level : 1
       user config file : /Users/piotrsobolewski/.condarc
 populated config files : /Users/piotrsobolewski/Dev/miniforge3/.condarc
                          /Users/piotrsobolewski/.condarc
          conda version : 22.9.0
    conda-build version : not installed
         python version : 3.10.2.final.0
       virtual packages : __osx=12.6=0
                          __unix=0=0
                          __archspec=1=arm64
       base environment : /Users/piotrsobolewski/Dev/miniforge3  (writable)
      conda av data dir : /Users/piotrsobolewski/Dev/miniforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /Users/piotrsobolewski/Dev/miniforge3/pkgs
                          /Users/piotrsobolewski/.conda/pkgs
       envs directories : /Users/piotrsobolewski/Dev/miniforge3/envs
                          /Users/piotrsobolewski/.conda/envs
               platform : osx-arm64
             user-agent : conda/22.9.0 requests/2.27.1 CPython/3.10.2 Darwin/21.6.0 OSX/12.6
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

**Comment by wolfv** [comment]
Does this file exist? 

"/Users/piotrsobolewski/Dev/miniforge3/etc/fish/conf.d/mamba.fish"

**Comment by psobolewskiPhD** [comment]
yes sir:
```
╭─    ~/Dev/miniforge3/etc/fish/conf.d ······························· ✔ ─
╰─ ll                                                                  (base) ─╯
.rw-rw-r-- 4.0k piotrsobolewski  1 Nov 22:00 conda.fish
.rw-rw-r--  690 piotrsobolewski  2 Nov 11:40 mamba.fish

```
here's the contents:
```
   1   │ # Based on conda.fish which is licensed as below
   2   │ #
   3   │ # Copyright (C) 2012 Anaconda, Inc
   4   │ # SPDX-License-Identifier: BSD-3-Clause
   5   │ 
   6   │ set -gx MAMBA_EXE (dirname $CONDA_EXE)/mamba
   7   │ 
   8   │ function mamba --inherit-variable CONDA_EXE --inherit-variable MAMBA_EX
       │ E
   9   │   if test (count $argv) -lt 1 || contains -- --help $argv
  10   │     $MAMBA_EXE $argv
  11   │   else
  12   │     set -l cmd $argv[1]
  13   │     set -e argv[1]
  14   │     switch $cmd
  15   │       case activate deactivate
  16   │         eval ($CONDA_EXE shell -s fish $cmd $argv)
  17   │       case install update upgrade remove uninstall
  18   │         $MAMBA_EXE $cmd $argv || return $status
  19   │         and eval ($CONDA_EXE shell -s fish reactivate)
  20   │       case '*'
  21   │         $MAMBA_EXE $cmd $argv
  22   │     end
  23   │   end
  24   │ end
```

**Comment by psobolewskiPhD** [comment]
> Argh, I think this needs to be
> 
> ```
> $CONDA_EXE shell.fish reactivate
> ```

Bingo.

**Comment by wolfv** [comment]
Would you be able to send a PR? We can make a follow up release quite shortly :)

**Comment by dlukes** [comment]
In that case, it probably also needs to be changed here: https://github.com/mamba-org/mamba/blob/679c13b232a33c5f2c7f6a62b9de22fae921ae89/libmamba/data/mamba.fish#L57-L73

Which is where I lifted it from.

I did notice the `shell -s fish` vs. `shell.fish` difference when preparing the PR (personally, I've been using `shell.fish` in my workaround, cf. [comment above](https://github.com/mamba-org/mamba/issues/1718#issuecomment-1271657234)), but I thought they were equivalent and figured I should stick with the same approach that's already used by micromamba in the same repo.

Apologies that I didn't uncover this while testing the PR, I thought I'd deactivated my own workaround, but probably not.

Should I open a PR which amends both `mamba/mamba/shell_templates/mamba.fish` and `libmamba/data/mamba.fish`? Or will a core team member just push a quick fix?
